src/cancelation_token.rs

```diff
@@ -0,0 +1,60 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+#[derive(Debug)]
+pub struct CancelationToken {
+	shared_state: Arc<Mutex<CancelationTokenState>>
+}
+
+#[derive(Debug)]
+struct CancelationTokenState {
+	canceled: bool,
+	waker: Option<Waker>
+}
+
+/// Future that allows gracefully shutting down the server
+impl CancelationToken {
+	pub fn new() -> CancelationToken {
+		CancelationToken {
+			shared_state: Arc::new(Mutex::new(CancelationTokenState {
+				canceled: false,
+				waker: None
+			}))
+		}
+	}
+
+	/// Call to shut down the server
+	pub fn complete(&self) {
+		let mut shared_state = self.shared_state.lock().unwrap();
+
+		shared_state.canceled = true;
+		if let Some(waker) = shared_state.waker.take() {
+			waker.wake()
+		}
+	}
+}
+
+impl Future for CancelationToken {
+	type Output = ();
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		let mut shared_state = self.shared_state.lock().unwrap();
+
+		if shared_state.canceled {
+            Poll::Ready(())
+		} else {
+            shared_state.waker = Some(cx.waker().clone());
+            Poll::Pending
+		}
+	}
+}
+
+impl Clone for CancelationToken {
+	fn clone(&self) -> Self {
+		CancelationToken {
+			shared_state: self.shared_state.clone()
+		}
+	}
+}
```

Comment:
The [stop-token](https://docs.rs/crate/stop-token/0.1.2) does much the same, and we're looking to integrate it into async-std in the near future. If possible it'd be great to use that crate directly instead of defining our own here.

Reply:
One thing I like about stop-token is that it supports streams.

However, the authors describe it as "experimental." Upon closer inspection:

- The documentation doesn't build: https://docs.rs/crate/stop-token/0.1.2
- The tests are limited: https://github.com/async-rs/stop-token/blob/master/tests/tests.rs
- Based on how I interpret https://github.com/async-rs/stop-token/blob/master/src/lib.rs#L170, canceling via stop token cancels the next read from the stream. (IE, if I cancel NOW, the background task doesn't cancel until there's another item in the stream.)

I should point out that the library that I made with cancelation token has full documentation and tests: https://crates.io/crates/sync-tokens